feat: Add session recordings

holloway opened the pull request:
Replaces #8084

* Allows a secretariat to add and delete session recordings (youtube urls)

![Screenshot from 2024-11-18 11-11-17](https://github.com/user-attachments/assets/8304b703-54be-4c63-8b81-5ce9ea344f06)

![Screenshot from 2024-11-18 11-11-37](https://github.com/user-attachments/assets/cb2ebca5-a9f2-466c-bb38-42671d1a4f33)

holloway committed:
feat: add session recordings

holloway committed:
feat: add session recordings

holloway committed:
feat: deleting recordings

holloway committed:
feat: deleting recordings and initial form values

holloway commented on ietf/meeting/views.py:
```diff
@@ -2568,6 +2569,66 @@ def add_session_drafts(request, session_id, num):
+
+    if request.method == 'POST':
+        delete = request.POST['delete']
+        if delete:
+            delete_recording(pk=delete, session=session)
```
>  We should think through what a chair would do if they entered the wrong link and wanted to fix it. @rjsparks 

I've added a delete button. Is that ok?


![Screenshot from 2024-11-18 10-56-26](https://github.com/user-attachments/assets/82295bc2-73b2-46c6-9a0b-864db4d140e4)

holloway replied:
Regarding the 'Revision' column is that valid for recordings? How would someone update the revision of a recording anyway?

rjsparks replied:
It's valid as long as we're modeling recordings as Document objects.

To date, we have not revised a recording document. Doesn't mean we couldn't (with the current model anyhow). An -01 for youtube would just point to a different URL the same as an -01 for a draft points to a different filename.

holloway replied:
Thoughts on the delete button? Should it have a confirmation modal?

holloway replied:
I'm going to assume we need a confirmation modal.

rjsparks replied:
I have minor heartburn that we're allowing a Document object to be deleted. It violates an assumption about what Document objects are _for_, but I suppose we have that problem with these recording documents-that-are-just-external-links already and some future project should migrate them to some other model instead.
 
I'll request one change and we can move forward with this.

jennifer-richards commented on ietf/meeting/views.py:
```diff
@@ -2568,6 +2569,66 @@ def add_session_drafts(request, session_id, num):
+                return redirect('ietf.meeting.views.session_details', num=session.meeting.number, acronym=session.group.acronym)
+    
+    today = datetime.datetime.now()
+    initial = {'title': f"Video recording for {session.group.acronym} on {today.strftime('%b-%d-%Y at %H:%M:%S')}"}
```
I don't know what the existing TZ is, but I think UTC is the right choice. I'd probably add "... UTC" to the default name, though it'd mean a break from existing filename convention.

rjsparks replied:
Why today?

The recording is for a working group session and the times should be the times for that session, not the time of the upload of the recording.

We should consider capturing the moment of "upload"  (really, setting a URL) as a docevent, but I don't think we're currently doing that when meetecho reports uploads to youtube, so maybe this is a separate issue for improving both.

But yes, we should present in UTC by default - some views offer browser or meeting local time as well - this upload doesn't need to though, I think.

holloway replied:
Done, it now looks like this

![Screenshot from 2024-11-19 09-51-01](https://github.com/user-attachments/assets/ca5da031-200e-464b-9d30-3c803a2c51fa)

holloway replied:
Based on the assumption that `session.meeting.date` is already UTC. Is that correct?

jennifer-richards replied:
Sadly, I don't think so, it'll be the date in the meeting timezone.

jennifer-richards replied:
`Meeting.start_datetime()` will give you a timezone-aware midnight in the meeting timezone. If you have a `Session` (which I think you do), then the way to get the start time of that session is:

jennifer-richards replied:
Also, this needs a guard against `official_timeslotassignment()` not existing unless it's guaranteed to exist. It should exist for anything that is on an agenda.

holloway committed:
feat: use meeting date rather than today for initial title field. Fix delete recording

holloway committed:
feat: confirm delete recordings modal. fix server utils delete recording

holloway committed:
fix: removing debug console.log

holloway committed:
feat: change button name from 'Ok' to 'Delete' for confirm deletion to be clearer

jennifer-richards commented on ietf/meeting/views.py:
```diff
@@ -2568,6 +2569,56 @@ def add_session_drafts(request, session_id, num):
+
+    session_number = None
+    sessions = get_sessions(session.meeting.number,session.group.acronym)
+    initial = {
+        'title': f"Video recording for {session.group.acronym} on {session.meeting.date.strftime('%b-%d-%Y at %H:%M:%S')}"
+    }
```
To be more concrete about my comments elsewhere, I think something like this is what is needed. The `assertion()` call is not bulletproof, there'll still be a 500 if this comes up.
```suggestion
    timeslot = session.official_timeslotassignment()
    assertion("timeslot is not None")
    initial = {
        'title': f"Video recording for {session.group.acronym} on {timeslot.utc_start_time().strftime('%b-%d-%Y at %H:%M:%S')}"
    }
```

jennifer-richards replied:
Noticed I botched the `timeslotassignment.timeslot` step - glad you caught it!

rjsparks commented on ietf/templates/meeting/add_session_recordings.html:
```diff
@@ -0,0 +1,113 @@
+        </table>
+    </form>
+    <dialog id="delete_confirm_dialog">
+        <p>Really delete <a href="#" id="delete_confirm_link"></a>?</p>
```
Could we say `Really delete the link to` instead? We don't want to leave anyone with the impression that this will remove the recording at youtube.

holloway replied:
Ok, it looks like this now

![Screenshot from 2024-11-20 07-33-26](https://github.com/user-attachments/assets/219b9a0d-3e49-4f25-8b98-99ba14ac50d3)

holloway committed:
feat: UTC time in string and delete modal text

holloway committed:
fix: django html validation tests

holloway committed:
fix: django html validation tests

holloway committed:
fix: django html validation tests

rjsparks reviewed (changes requested):
I think this will work now (and it seems fine with manual testing). We need CI tests though.

rjsparks commented on ietf/meeting/utils.py:
```diff
@@ -870,6 +870,14 @@ def create_recording(session, url, title=None, user=None):
+    '''
+    document = Document.objects.filter(pk=pk, group=session.group).first()
+    if document:
+        document.delete()
```
This is what caught my eye. This function would be better accepting a Document object rather than a pk, and perhaps should live in the Session model so that the connection to Session is naturally enforced.

jennifer-richards commented:
This is a suggestion following up @rjsparks' comments inline regarding deletion of the `Document` and the interface for doing so. The inline conversation was long enough I think this might be easier to follow.

I think the "Delete the link to..." action should delete the `SessionPresentation` that links to the `Document` rather than the `Document` itself. This avoids the worry about changing the historical policy about `Document` lifetimes. It also lets the `delete_recording()` action restrict itself to the `Session` rather than using `session.group` as a proxy and trusting the discretion/awareness of materials managers.

When deleting the `SessionPresentation`, the `Document` should have its "recording" state set to "deleted" rather than actually being deleted from the database.

As far as pk vs instances, I don't think the modeling will let us use `ModelForm` in the way @rjsparks was suggesting. If it can, that'd be better, but `SessionPresentation` is really just an annotated through model between `Document` and `Session`. If this was planned to be a traditional HTML forms interface, the right approach would be a `FormSet` that could do that magic, but as an endpoint for a JS front end I think we have to roll our own.

rjsparks commented:
This particular view isn't using a js frontend.

jennifer-richards commented:
> This particular view isn't using a js frontend.

Ah, I guess I'd mixed it up with work in other PRs. In that case we might want to use a formset

jennifer-richards committed:
Merge branch 'main' into add-recordings-mk2

jennifer-richards committed:
refactor: Work with SessionPresentations

jennifer-richards committed:
fix: better ordering

jennifer-richards committed:
chore: drop rev, hide table when empty

jennifer-richards committed:
test: test delete_recordings method

jennifer-richards committed:
fix: debug delete_recordings

holloway committed:
Merge branch 'ietf-tools:main' into add-recordings

jennifer-richards committed:
test: test add_session_recordings view

jennifer-richards committed:
fix: better permissions handling

jennifer-richards committed:
Merge remote-tracking branch 'holloway/add-recordings' into add-recordings-mk2

jennifer-richards committed:
fix: only delete recordings for selected session

jennifer-richards committed:
refactor: inline script -> js module

jennifer-richards committed:
chore: remove accidental import

*shakes fist at pycharm*

jennifer-richards committed:
fix: consistent timestamp format

plus slight rephrase

jennifer-richards committed:
style: Black

jennifer-richards committed:
chore: remove comment

jennifer-richards committed:
test: update test to match

jennifer-richards committed:
fix: reversible url pattern for materials

Tests were perturbed in a way that led to a test
getting an interim instead of an IETF meeting.
This exposed a bug reversing the URL for the
materials_document() view. This splits it into
two patterns that are equivalent to the original.
